### PR TITLE
add database and operational support for read groups in entity classes

### PR DESCRIPTION
**Pull request by pdowler**
caom2persistence:
- remove all reference to *ReadAccess classes
- add support for storing group URIs (Observation.metaReadGroups, Plane.metaReadGroups, Plane,.dataReadGroups)
- retain the optimised tsvector column with just groups names as-is
- automagic upgrade SQL for postgresql drops tables and obsolete stateCode column, adds text columns for List<URI>

caom2-access-control:
- moved access control generator to a library so it can be used in caom2-repo-server and caom2harvester
- changed to generate group list and put them directly into the Observation/Plane(s)

caom2harvester:
- did some rework so the --compute and --generate-ac options can be used with source-destination so we have a tool to redo the computations that can be enabled in caom2-repo-server; this replaces the caom2ac discovery agent since we no longer need to cleanup public grants

### CODE REVIEW

File: caom2persistence/src/intTest/java/ca/nrc/cadc/caom2/persistence/AbstractDeletedEntityDAOTest.java [hunk 1]
```diff
@@ -125,87 +114,16 @@ public AbstractDeletedEntityDAOTest(Class genClass, String server, String databa
         }
     }
 
-    @Before
+    //@Before
```
Comment (andamian) [inline caom2persistence/src/intTest/java/ca/nrc/cadc/caom2/persistence/AbstractDeletedEntityDAOTest.java]: Not needed?

File: caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java [hunk 2]
```diff
@@ -532,48 +527,48 @@ protected void init() {
         }
         columnMap.put(Chunk.class, chunkColumns);
 
-        String[] metaReadAccessCols = new String[]{
-            "assetID", "groupID", "lastModified", "stateCode", "metaChecksum", "readAccessID"
-        };
-        columnMap.put(ObservationMetaReadAccess.class, metaReadAccessCols);
-        columnMap.put(PlaneMetaReadAccess.class, metaReadAccessCols);
-        columnMap.put(PlaneDataReadAccess.class, metaReadAccessCols);
+        //String[] metaReadAccessCols = new String[]{
```
Comment (andamian) [inline caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java]: cleanup

File: caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java [hunk 3]
```diff
@@ -532,48 +527,48 @@ protected void init() {
...
-        columnMap.put(PlaneSkeleton.class, new String[]{"lastModified", "maxLastModified", "stateCode", "metaChecksum", "accMetaChecksum", "planeID"});
-        columnMap.put(ArtifactSkeleton.class, new String[]{"lastModified", "maxLastModified", "stateCode", "metaChecksum", "accMetaChecksum", "artifactID"});
-        columnMap.put(PartSkeleton.class, new String[]{"lastModified", "maxLastModified", "stateCode", "metaChecksum", "accMetaChecksum", "partID"});
-        columnMap.put(ChunkSkeleton.class, new String[]{"lastModified", "maxLastModified", "stateCode", "metaChecksum", "accMetaChecksum", "chunkID"});
-
-        columnMap.put(ObservationMetaReadAccessSkeleton.class, new String[]{"lastModified", "stateCode", "metaChecksum", "readAccessID"});
-        columnMap.put(PlaneMetaReadAccessSkeleton.class, new String[]{"lastModified", "stateCode", "metaChecksum", "readAccessID"});
-        columnMap.put(PlaneDataReadAccessSkeleton.class, new String[]{"lastModified", "stateCode", "metaChecksum", "readAccessID"});
+        //String[] deletedEntityCols = new String[]{
```
Comment (andamian) [inline caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java]: cleanup

File: caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java [hunk 4]
```diff
@@ -532,48 +527,48 @@ protected void init() {
...
+        //columnMap.put(DeletedPlaneMetaReadAccess.class, deletedEntityCols);
+        //columnMap.put(DeletedPlaneDataReadAccess.class, deletedEntityCols);
+
+        columnMap.put(ObservationSkeleton.class, new String[]{"lastModified", "maxLastModified", "metaChecksum", "accMetaChecksum", "obsID"});
+        columnMap.put(PlaneSkeleton.class, new String[]{"lastModified", "maxLastModified", "metaChecksum", "accMetaChecksum", "planeID"});
+        columnMap.put(ArtifactSkeleton.class, new String[]{"lastModified", "maxLastModified", "metaChecksum", "accMetaChecksum", "artifactID"});
+        columnMap.put(PartSkeleton.class, new String[]{"lastModified", "maxLastModified", "metaChecksum", "accMetaChecksum", "partID"});
+        columnMap.put(ChunkSkeleton.class, new String[]{"lastModified", "maxLastModified", "metaChecksum", "accMetaChecksum", "chunkID"});
+
+        //columnMap.put(ObservationMetaReadAccessSkeleton.class, new String[]{"lastModified", "metaChecksum", "readAccessID"});
```
Comment (andamian) [inline caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java]: cleanup

File: caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java [hunk 5]
```diff
@@ -985,17 +968,17 @@ protected String getUpdateChildOptimisationSQL(Class child) {
         return sb.toString();
     }
 
-    protected String getUpdateAssetSQL(Class asset, Class ra, boolean add) {
-        throw new UnsupportedOperationException();
-    }
-
+    //protected String getUpdateAssetSQL(Class asset, Class ra, boolean add) {
```
Comment (andamian) [inline caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java]: cleanup?

File: caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java [hunk 6]
```diff
@@ -3459,8 +3186,8 @@ public Observation mapRow(ResultSet rs, int row, int col)
             Util.assignLastModified(o, lastModified, "lastModified");
             Util.assignLastModified(o, maxLastModified, "maxLastModified");
 
-            Integer stateCode = Util.getInteger(rs, col++);
-            log.debug("found: observation.stateCode = " + stateCode);
+            //Integer stateCode = Util.getInteger(rs, col++);
```
Comment (andamian) [inline caom2persistence/src/main/java/ca/nrc/cadc/caom2/persistence/SQLGenerator.java]: cleanup here and below

File: caom2-access-control/src/main/java/ca/nrc/cadc/caom2/ac/ReadAccessGenerator.java [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+                    throw new RuntimeException("group not found: " + proposalGroupName + " for update (right after check/create)");
+                } catch (Exception e) {
+                    throw new RuntimeException("could not update group " + proposalGroupName, e);
+                }
+                log.debug("added group admins to group: " + proposalGroupName);
+            }
+
+            // cache groups we have already updated
+            updatedProposalGroups.add(proposalGroupName);
+            if (updatedProposalGroups.size() > 1000) {
```
[the remark this answers is not in the document]
Comment (pdowler) [reply caom2-access-control/src/main/java/ca/nrc/cadc/caom2/ac/ReadAccessGenerator.java]: That was from use in caom2ac and could still be useful if we run caom2harvester to regenerate.

File: caom2persistence/src/intTest/java/ca/nrc/cadc/caom2/persistence/AbstractDeletedEntityDAOTest.java (hunk 1 above)
Comment (pdowler) [reply caom2persistence/src/intTest/java/ca/nrc/cadc/caom2/persistence/AbstractDeletedEntityDAOTest.java]: cleanup not currently needed but was copied from another test class

**Pull request merged**